WIP: more enum layout optimizations

mikebenfield opened the pull request:
Work in progress, but it works. 

In addition to addressing #101567, also keep around multiple niches rather than just the largest one, and for tagged variants possibly move fields around the niche rather than just putting the whole variant before or after.

Things that aren't right at the moment:

- [ ] I haven't fixed cranelift codegen
- [ ] I'm just emitting bogus debuginfo in `rustc_codegen_llvm/src/debuginfo/metadata/enums/mod.rs`
- [X] Contrary to my comment in #101567, I only look at existing niches when creating a `Flag`, even though any field will do
- [ ] Keeping a `Vec` of niches rather than just the largest one likely has a significant negative effect on performance, but I've done no performance testing
- [ ] Comments should be updated/added to explain what's going on with `Flag`
- [ ] The current sorting of niches in the presence of `Flag` probably makes no sense
- [ ] There probably should be some `repr` to force an `enum` to use a `Flag`. Right now an enum like the one given in the issue:
```
pub enum Provenance {
    Concrete {
        alloc_id: NonZeroU64,
        sb: NonZeroU64,
    },
    Wildcard,
}
```
will just use a regular niche, which means `Option<Provenance>` won't be able to use a flag.

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3055469463/jobs/4928561297) [(plain)](https://github.com/rust-lang/rust/commit/a3308c16a0c907d0c861468b0b80678eaff65712/checks/8358385678/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 24466cae975a
Successfully tagged rust-ci:latest
Built container sha256:24466cae975af64876ac757dbf31ebf19f47edd8075ff919005aa2f8eab5b3a0
Uploading finished image to https://ci-caches.rust-lang.org/docker/100443de95ac5fd5d76f947e37204434fbf3cf3da99f888ec8bb4fbbceda7ed7d93b519998f8388225735159bb6cbee56e76ad72dc699a1cf736f3af809f03d0
upload failed: - to s3://rust-lang-ci-sccache2/docker/100443de95ac5fd5d76f947e37204434fbf3cf3da99f888ec8bb4fbbceda7ed7d93b519998f8388225735159bb6cbee56e76ad72dc699a1cf736f3af809f03d0 Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0027]: pattern does not mention field `flag`
  --> src/discriminant.rs:45:27
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start },
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
   |
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag },
   |                                                                                                 ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. },

error[E0027]: pattern does not mention field `flag`
   --> src/discriminant.rs:116:9
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start } => {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag } => {
    |                                                                               ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. } => {

For more information about this error, try `rustc --explain E0027`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:03:21
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #98588) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3075030393/jobs/4968167478) [(plain)](https://github.com/rust-lang/rust/commit/a85290cc7304ed50e41601421bed2eb548afe1da/checks/8410785412/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0027]: pattern does not mention field `flag`
  --> src/discriminant.rs:45:27
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start },
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
   |
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag },
   |                                                                                                 ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. },

error[E0027]: pattern does not mention field `flag`
   --> src/discriminant.rs:116:9
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start } => {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag } => {
    |                                                                               ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. } => {

For more information about this error, try `rustc --explain E0027`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:02:51
```

</details>

RalfJung commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -730,14 +730,35 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                        Err(_) => {
+                            // So this is a pointer then, and casting to an int failed. Can only
+                            // happen during CTFE.
+                            // The flag is not set to magic_value
```
This is suspicious. We should probably abort execution here, I am fairly sure just silently continuing with a fallback will produce incorrect results.

mikebenfield replied:
I'm not sure this is accurate. If the address is relative (`Provenance::OFFSET_IS_ADDR` is `false`), we still should be able to read the discriminant without just aborting. Essentially the same thing happens below already, in computing the discriminant for a regular old niche.

RalfJung replied:
But here you just assume that the pointer (when converted to an integer) is not equal to the magic value. What is the basis for that assumption?

mikebenfield replied:
> But here you just assume that the pointer (when converted to an integer) is not equal to the magic value. What is the basis for that assumption?

The only possible magic value for a pointer is 0, and if the pointer fails to convert to an integer, it's not null. I will add a similar check to what appears below, but my understanding is that the check below is like a sanity check or assertion, not checking for something that is intended to occur in practice.

RalfJung replied:
Its' not an `assert!` because it can definitely happen in practice. You just need to write the right kind of `unsafe const` code to trigger it.

RalfJung replied:
[Here's a way to trigger that error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f25cdb7cd38d79be3d5370a55d8803c9).

RalfJung replied:
>  if the pointer fails to convert to an integer, it's not null. 

That assumption is wrong. `(&0 as *const i32).wrapping_offset(0x10000)` will fail to convert to an integer, but might be null (depending on where that `0` will be allocated at runtime).

RalfJung commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -730,14 +730,35 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
-                    Err(dbg_val) => {
+                let variant = match (is_magic_val, tag_val.try_to_int()) {
+                    (false, _) => untagged_variant,
+                    (_, Err(dbg_val)) => {
```
```suggestion
                    (true, Err(dbg_val)) => {
```

RalfJung commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -750,7 +771,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                         untagged_variant
                     }
-                    Ok(tag_bits) => {
+                    (_, Ok(tag_bits)) => {
```
```suggestion
                    (true, Ok(tag_bits)) => {
```
explicit is better than implicit

mikebenfield commented:
- Fixed issues pointed out by @RalfJung 
- Use `SmallVec` in a couple places
- better sorting (maybe) of niches
- better search for a niche when using a flag

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3078953067/jobs/4974828533) [(plain)](https://github.com/rust-lang/rust/commit/5155af905d1deea04d7c4600a2f2bd8834910c51/checks/8418618091/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0027]: pattern does not mention field `flag`
  --> src/discriminant.rs:45:27
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start },
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
   |
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag },
   |                                                                                                 ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. },

error[E0027]: pattern does not mention field `flag`
   --> src/discriminant.rs:116:9
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start } => {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag } => {
    |                                                                               ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. } => {

For more information about this error, try `rustc --explain E0027`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:02:47
```

</details>

mikebenfield commented:
I implemented `repr(flag)`, which can go on an enum to indicate that we should try a niche with a new flag _first_, rather than after looking at existing niches. I also added `Option<Provenance>` from #101567 as a test case.

mikebenfield commented:
About `repr(flag)`:

- its name probably should change;
- I'm not sure it should exist;
- I assume it should go through an RFC process or something.

RalfJung commented on compiler/rustc_const_eval/src/interpret/validity.rs:
```diff
@@ -220,8 +220,16 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
+                } else {
+                    None
+                };
+                if tag_field == field || flag_field == Some(field) {
```
```suggestion
                // For now, let's just report both tag and flag fields as being part of the tag.
                if tag_field == field || flag_field == Some(field) {
```

RalfJung commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -730,14 +730,43 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                            // So this is a pointer then, and casting to an int
+                            // failed. Can only happen during CTFE. If the magic
+                            // value is 0 and the scalar is not null, we know
+                            // this is OK. Anything else we conservatively
+                            // reject.
+                            let ptr_valid =
```
```suggestion
                            // the pointer cannot be the magic value. Anything else
                            // we conservatively reject.
                            let ptr_definitely_not_magic_val =
```
plus renaming the variable in its two uses below

(`ptr_valid` is a bad name also in the later code for niche handling, but it's much more confusing here IMO. if you want to rename it to `ptr_definitely_not_in_niche_variants` that would be great)

RalfJung commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -739,11 +739,11 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
-                            let ptr_valid =
+                            // the pointer cannot be the magic value. Anything
+                            // else we conservatively reject.
+                            let ptr_definitely_not_in_niche_variants =
```
Nono this is the flag one, not the niche variants one. ;)

RalfJung replied:
This should be ptr_definitely_not_magic_val.

The *other* one should be ptr_definitely_not_in_niche_variants.

mikebenfield replied:
oh oops!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3081009829/jobs/4979053972) [(plain)](https://github.com/rust-lang/rust/commit/b165ea6a06b9ed50860c63ab00f615a2d68daffc/checks/8423948578/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0027]: pattern does not mention field `flag`
  --> src/discriminant.rs:45:27
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start },
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
   |
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag },
   |                                                                                                 ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. },

error[E0027]: pattern does not mention field `flag`
   --> src/discriminant.rs:116:9
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start } => {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag } => {
    |                                                                               ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. } => {

For more information about this error, try `rustc --explain E0027`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:02:49
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3083172914/jobs/4983778381) [(plain)](https://github.com/rust-lang/rust/commit/880d4931961581cbcd9dc50211c9459f2541b2dd/checks/8430133046/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0027]: pattern does not mention field `flag`
  --> src/discriminant.rs:45:27
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start },
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
   |
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag },
   |                                                                                                 ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. },

error[E0027]: pattern does not mention field `flag`
   --> src/discriminant.rs:116:9
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start } => {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag } => {
    |                                                                               ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. } => {

For more information about this error, try `rustc --explain E0027`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:03:30
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #102051) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

mikebenfield commented:
Status: In working on this, I've encountered what I'm 95% sure is a bug in the LLVM InstCombine pass that manifests in some cases with these layout optimizations. I'm going to investigate more, and then either fix that or report it, then will return to this PR.

Commit:
Use a Flag to enable more niche optimizations.

Also:
Try to fit other variants by moving fields around the niche.
Keep multiple niches, not just the largest one.
Look for multiple largest variants.
Introduce repr(flag).

Fixes #101567

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3201625260/jobs/5229795123) [(plain)](https://github.com/rust-lang/rust/commit/3ea0d8f9977126cb0f17f80b5cb1cb65d9374248/checks/8756467872/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_cranelift v0.1.0 (/checkout/compiler/rustc_codegen_cranelift)
error[E0027]: pattern does not mention field `flag`
  --> src/discriminant.rs:45:27
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start },
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
   |
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag },
   |                                                                                                 ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
45 |             tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. },

error[E0027]: pattern does not mention field `flag`
   --> src/discriminant.rs:116:9
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start } => {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `flag`
help: include the missing field in the pattern
    |
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, flag } => {
    |                                                                               ~~~~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
    |
116 |         TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. } => {

For more information about this error, try `rustc --explain E0027`.
error: could not compile `rustc_codegen_cranelift` due to 2 previous errors
Build completed unsuccessfully in 0:04:01
```

</details>

mikebenfield commented:
Rebased. Since some things were moved around it was most expedient to just squash the commits.

I fixed the problem I mentioned above, which was in fact not an LLVM bug. The issue is that when there's a flag, the values of the niche are only actually restricted to its valid_range when the flag is set to its magic value. The solution I've adopted for now is, when there's a flag, to just emit the full range of the primitive as the valid range (see `rustc_middle/src/ty/layout.rs`). An alternative  would be, when there's a flag, only access the niche when the flag is set to magic_value. This would a couple extra branches in computing the discriminant, although maybe LLVM could optimize those branches away. I'll experiment at some point.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3201664924/jobs/5229877970) [(plain)](https://github.com/rust-lang/rust/commit/6457935328a5301fff890a76c2cf6f335fab8130/checks/8756573009/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_codegen_cranelift/src/abi/comments.rs at line 82:
     }
     }
     let TyAndLayout { ty, layout } = place.layout();
-    let rustc_target::abi::LayoutS {
-        align,
-        abi: _,
-        variants: _,
-        fields: _,
-        fields: _,
-        niches: _,
-    } = layout.0.0;
+    let rustc_target::abi::LayoutS { size, align, abi: _, variants: _, fields: _, niches: _ } =
+        layout.0.0;
 
     let (kind, extra) = match *place.inner() {
         CPlaceInner::Var(place_local, var) => {
Diff in /checkout/compiler/rustc_codegen_cranelift/src/discriminant.rs at line 42:
         Variants::Multiple {
             tag: _,
             tag_field,
-            tag_encoding: TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. },
+            tag_encoding:
+                TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start, .. },
             variants: _,
         } => {
             if variant_index != untagged_variant {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_cranelift/src/abi/comments.rs" "/checkout/compiler/rustc_codegen_cranelift/src/vtable.rs" "/checkout/compiler/rustc_codegen_cranelift/src/lib.rs" "/checkout/compiler/rustc_codegen_cranelift/src/abi/returning.rs" "/checkout/compiler/rustc_codegen_cranelift/src/abi/mod.rs" "/checkout/compiler/rustc_codegen_cranelift/src/abi/pass_mode.rs" "/checkout/compiler/rustc_codegen_cranelift/src/optimize/peephole.rs" "/checkout/compiler/rustc_codegen_cranelift/src/main_shim.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3201754419/jobs/5230042085) [(plain)](https://github.com/rust-lang/rust/commit/418c2788bd5e583d05e9d85266e0b3f44adeb7bb/checks/8756783973/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_target/src/abi/mod.rs at line 1195:
     ) -> Option<Self> {
         let niche_data = |offset, scalar| {
             let Scalar::Initialized {value, valid_range } = scalar else { return None };
-            Some(NicheData {
-                value,
-                valid_range,
-            })
-            })
+            Some(NicheData { offset, value, valid_range })
 
 
         let niche = Niche {
Diff in /checkout/compiler/rustc_target/src/abi/mod.rs at line 1206:
             data: niche_data(niche_offset, niche_scalar)?,
-            flag: Some(
-                niche_data(flag_offset, flag_scalar)?,
-            )
+            flag: Some(niche_data(flag_offset, flag_scalar)?),
 
 
         if niche.available(cx) > 0 || niche.available_with_new_flag(cx) > 0 {
Diff in /checkout/compiler/rustc_target/src/abi/mod.rs at line 1237:
                 let niche_size = self.data.value.size(cx);
                 assert!(niche_size.bits() <= 128);
                 let max = size.unsigned_int_max();
-                if max < u128::MAX {
-                    max + 1
-                    max
-                }
-                }
+                if max < u128::MAX { max + 1 } else { max }
                 0
             }
Diff in /checkout/compiler/rustc_target/src/abi/mod.rs at line 1248:
         })
         })
     }
 
-    pub fn reserve_with_new_flag<C: HasDataLayout>(&self, cx: &C, count: u128) -> Option<(u128, Scalar, Scalar)> {
+    pub fn reserve_with_new_flag<C: HasDataLayout>(
+        &self,
+        cx: &C,
+        count: u128,
+    ) -> Option<(u128, Scalar, Scalar)> {
         assert!(count > 0);
 
         self.flag.map_or(None, |flag_data| {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_ast/src/token.rs" "/checkout/compiler/rustc_llvm/src/lib.rs" "/checkout/compiler/rustc_llvm/build.rs" "/checkout/compiler/rustc_ast/src/util/comments.rs" "/checkout/compiler/rustc_traits/src/lib.rs" "/checkout/compiler/rustc_target/src/json.rs" "/checkout/compiler/rustc_target/src/abi/mod.rs" "/checkout/compiler/rustc_ast/src/entry.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #102875) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jackh726 commented:
Do you want a review of this or are you still working on it?

mikebenfield commented:
Let's hold off on a review for now. I'm going to modify it to fix #102997 also.

JohnCSimon commented:
@mikebenfield 
ping from triage - can you post your status on this PR? Thanks

mikebenfield commented:
@JohnCSimon I've got local in-progress changes that are semi-close to being done. I've been meaning to get back to this anyway so I'll see if I can get them done and pushed here today or tomorrow. Can't guarantee it'll be done though.

mikebenfield commented:
Status: I think I'll have this ready for review by end of week.

I had it pretty much done and then got laid off and lost access to the code I had written 😂 . So I kind of had to start from scratch, but it should get done quickly this time.

Dylan-DPC commented:
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
